File: sybau_frontend/src/utils/authValidation.ts
export const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/i;

const disposableEmailDomains = new Set([
  '10minutemail.com',
  'dispostable.com',
  'emailondeck.com',
  'fakeinbox.com',
  'getnada.com',
  'grr.la',
  'guerrillamail.com',
  'mail.tm',
  'maildrop.cc',
  'mailinator.com',
  'moakt.com',
  'mytemp.email',
  'sharklasers.com',
  'tempmail.com',
  'temp-mail.org',
  'tempr.email',
  'throwawaymail.com',
  'tmail.io',
  'trashmail.com',
  'yopmail.com',
]);

export const passwordRequirements = [
  { key: 'length', label: 'Mindestens 8 Zeichen', test: (value: string) => value.length >= 8 },
  { key: 'lower', label: 'Ein Kleinbuchstabe', test: (value: string) => /[a-z]/.test(value) },
  { key: 'upper', label: 'Ein Großbuchstabe', test: (value: string) => /[A-Z]/.test(value) },
  { key: 'number', label: 'Eine Zahl', test: (value: string) => /\d/.test(value) },
  { key: 'special', label: 'Ein Sonderzeichen', test: (value: string) => /[^A-Za-z0-9]/.test(value) },
];

export function normalizeEmail(value: string) {
  return value.trim().toLowerCase();
}

export function isDisposableEmail(value: string) {
  const domain = normalizeEmail(value).split('@')[1] ?? '';
  return [...disposableEmailDomains].some((blockedDomain) =>
    domain === blockedDomain || domain.endsWith(`.${blockedDomain}`),
  );
}

export function validateEmail(value: string, options: { required?: boolean; blockDisposable?: boolean } = {}) {
  const email = normalizeEmail(value);
  if (!email) return options.required === false ? '' : 'Bitte gib eine E-Mail-Adresse ein.';
  if (!emailRegex.test(email)) return 'Bitte gib eine gültige E-Mail-Adresse ein.';
  if (options.blockDisposable && isDisposableEmail(email)) {
    return 'Temporäre E-Mail-Adressen sind nicht erlaubt.';
  }
  return '';
}

export function getPasswordRequirementStates(value: string) {
  return passwordRequirements.map((requirement) => ({
    ...requirement,
    valid: requirement.test(value),
  }));
}

export function validatePassword(
  value: string,
  options: { required?: boolean; requireStrength?: boolean } = {},
) {
  if (!value) return options.required === false ? '' : 'Bitte gib ein Passwort ein.';
  if (!options.requireStrength) return '';

  const failedRequirement = passwordRequirements.find((requirement) => !requirement.test(value));
  if (failedRequirement) return 'Das Passwort erfüllt noch nicht alle Anforderungen.';
  if (value.length > 128) return 'Das Passwort darf maximal 128 Zeichen lang sein.';
  return '';
}

File: Sybau_Backend/Sybau_Backend/Validation/AuthValidation.cs
using System.ComponentModel.DataAnnotations;
using System.Text.RegularExpressions;

namespace Sybau_Backend.Validation;

public static partial class AuthValidation
{
    public const int MinPasswordLength = 8;
    public const int MaxPasswordLength = 128;
    public const string PasswordPolicyMessage =
        "Passwort muss mindestens 8 Zeichen lang sein und Großbuchstaben, Kleinbuchstaben, Zahl und Sonderzeichen enthalten.";

    private static readonly HashSet<string> DisposableEmailDomains = new(StringComparer.OrdinalIgnoreCase)
    {
        "10minutemail.com",
        "dispostable.com",
        "emailondeck.com",
        "fakeinbox.com",
        "getnada.com",
        "grr.la",
        "guerrillamail.com",
        "mail.tm",
        "maildrop.cc",
        "mailinator.com",
        "moakt.com",
        "mytemp.email",
        "sharklasers.com",
        "tempmail.com",
        "temp-mail.org",
        "tempr.email",
        "throwawaymail.com",
        "tmail.io",
        "trashmail.com",
        "yopmail.com"
    };

    public static string NormalizeEmail(string email) => email.Trim().ToLowerInvariant();

    public static bool HasValidEmailShape(string email)
    {
        var normalized = NormalizeEmail(email);
        return EmailRegex().IsMatch(normalized);
    }

    public static bool IsDisposableEmail(string email)
    {
        var domain = NormalizeEmail(email).Split('@').LastOrDefault() ?? string.Empty;
        return DisposableEmailDomains.Any(blocked =>
            string.Equals(domain, blocked, StringComparison.OrdinalIgnoreCase) ||
            domain.EndsWith($".{blocked}", StringComparison.OrdinalIgnoreCase));
    }

    public static bool IsStrongPassword(string password)
    {
        return password.Length is >= MinPasswordLength and <= MaxPasswordLength
            && password.Any(char.IsLower)
            && password.Any(char.IsUpper)
            && password.Any(char.IsDigit)
            && password.Any(ch => !char.IsLetterOrDigit(ch));
    }

    [GeneratedRegex(@"^[^\s@]+@[^\s@]+\.[^\s@]{2,}$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant)]
    private static partial Regex EmailRegex();
}

public sealed class NotDisposableEmailAttribute : ValidationAttribute
{
    public NotDisposableEmailAttribute()
    {
        ErrorMessage = "Temporäre E-Mail-Adressen sind nicht erlaubt.";
    }

    public override bool IsValid(object? value)
    {
        if (value is not string email || string.IsNullOrWhiteSpace(email)) return true;
        return AuthValidation.HasValidEmailShape(email) && !AuthValidation.IsDisposableEmail(email);
    }
}

public sealed class StrongPasswordAttribute : ValidationAttribute
{
    public StrongPasswordAttribute()
    {
        ErrorMessage = AuthValidation.PasswordPolicyMessage;
    }

    public override bool IsValid(object? value)
    {
        if (value is not string password || string.IsNullOrWhiteSpace(password)) return false;
        return AuthValidation.IsStrongPassword(password);
    }
}
